platform/darwin/test/MGLPredicateTests.mm

```diff
@@ -23,558 +23,404 @@ @interface MGLPredicateTests : XCTestCase
+        [self testSymmetryWithPredicate:predicate
+                          mustRoundTrip:NO];
     }
-
     {
-        mbgl::style::AnyFilter filter;
-        XCTAssertEqualObjects([NSPredicate mgl_predicateWithFilter:filter], [NSPredicate predicateWithValue:NO]);
+        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"NOT a IN {'b', 'c'}"];
+        NSArray *jsonExpression = @[@"!", @[@"match", @[@"get", @"a"], @[@"literal", @[@"b", @"c"]]]];
+        XCTAssertEqualObjects(predicate.mgl_jsonExpressionObject, jsonExpression);
+        XCTAssertEqualObjects([NSPredicate mgl_predicateWithJSONObject:jsonExpression], predicate);
+        [self testSymmetryWithPredicate:[NSPredicate mgl_predicateWithJSONObject:jsonExpression]
+                          mustRoundTrip:NO];
     }
-
     {
-        mbgl::style::AnyFilter filter = {
-            .filters = {
-                mbgl::style::EqualsFilter { .key = "a", .value = std::string("b") },
-                mbgl::style::EqualsFilter { .key = "c", .value = std::string("d") },
-            },
-        };
-        XCTAssertEqualObjects([NSPredicate mgl_predicateWithFilter:filter], [NSPredicate predicateWithFormat:@"a == 'b' OR c == 'd'"]);
+        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"NOT a IN %@", @[@"b", @"c"]];
+        NSArray *jsonExpression = @[@"!", @[@"match", @[@"get", @"a"], @[@"literal", @[@"b", @"c"]]]];
+        XCTAssertEqualObjects(predicate.mgl_jsonExpressionObject, jsonExpression);
+        XCTAssertEqualObjects([NSPredicate mgl_predicateWithJSONObject:jsonExpression], predicate);
+        [self testSymmetryWithPredicate:[NSPredicate mgl_predicateWithJSONObject:jsonExpression]
+                          mustRoundTrip:NO];
     }
-
     {
-        mbgl::style::NoneFilter filter;
-        XCTAssertEqualObjects([NSPredicate mgl_predicateWithFilter:filter], [NSPredicate predicateWithValue:YES]);
+        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"NOT {'b', 'c'} CONTAINS a"];
+        NSArray *jsonExpression = @[@"!", @[@"has", @[@"literal", @[@"b", @"c"]], @[@"get", @"a"]]];
+        XCTAssertEqualObjects(predicate.mgl_jsonExpressionObject, jsonExpression);
+        XCTAssertEqualObjects([NSPredicate mgl_predicateWithJSONObject:jsonExpression], predicate);
+        [self testSymmetryWithPredicate:[NSPredicate mgl_predicateWithJSONObject:jsonExpression]
+                          mustRoundTrip:NO];
     }
-
     {
-        mbgl::style::NoneFilter filter = {
-            .filters = {
-                mbgl::style::EqualsFilter { .key = "a", .value = std::string("b") },
-                mbgl::style::EqualsFilter { .key = "c", .value = std::string("d") },
-            },
-        };
-        XCTAssertEqualObjects([NSPredicate mgl_predicateWithFilter:filter], [NSPredicate predicateWithFormat:@"NOT(a == 'b' OR c == 'd')"]);
+        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"NOT %@ CONTAINS a", @[@"b", @"c"]];
```

Comment:
#11632 will also test using an aggregate expression instead of a constant array value.